### PR TITLE
[SPARK-23159][PYTHON] Update cloudpickle to v0.4.3

### PR DESCRIPTION
**Pull request by BryanCutler**
## What changes were proposed in this pull request?

The version of cloudpickle in PySpark was close to version 0.4.0 with some additional backported fixes and some minor additions for Spark related things.  This update removes Spark related changes and matches cloudpickle [v0.4.3](https://github.com/cloudpipe/cloudpickle/releases/tag/v0.4.3):

Changes by updating to 0.4.3 include:
* Fix pickling of named tuples https://github.com/cloudpipe/cloudpickle/pull/113
* Built in type constructors for PyPy compatibility [here](https://github.com/cloudpipe/cloudpickle/commit/d84980ccaafc7982a50d4e04064011f401f17d1b)
* Fix memoryview support https://github.com/cloudpipe/cloudpickle/pull/122
* Improved compatibility with other cloudpickle versions https://github.com/cloudpipe/cloudpickle/pull/128
* Several cleanups https://github.com/cloudpipe/cloudpickle/pull/121 and [here](https://github.com/cloudpipe/cloudpickle/commit/c91aaf110441991307f5097f950764079d0f9652)
* [MRG] Regression on pickling classes from the __main__ module https://github.com/cloudpipe/cloudpickle/pull/149
* BUG: Handle instance methods of builtin types https://github.com/cloudpipe/cloudpickle/pull/154
* Fix <span>#</span>129 : do not silence RuntimeError in dump() https://github.com/cloudpipe/cloudpickle/pull/153

## How was this patch tested?

Existing pyspark.tests using python 2.7.14, 3.5.2, 3.6.3

### CODE REVIEW

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: MAINT: Handle builtin type __new__ attrs: https://github.com/cloudpipe/cloudpickle/commit/f0d2011f9fc88105c174b7c861f2c2f56e870350

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: This exception handling is Spark specific, it has been moved to serializers.py `CloudPickleSerializer.dumps`

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: Some cleanups, fix memoryview support https://github.com/cloudpipe/cloudpickle/commit/f8187e90aed7e1b96ffaae85cdf4b37108c75d3f

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: BUG: Hit the builtin type cache for any function https://github.com/cloudpipe/cloudpickle/commit/d84980ccaafc7982a50d4e04064011f401f17d1b

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: BUG: Fix bug pickling namedtuple https://github.com/cloudpipe/cloudpickle/commit/28070bba79cf71e5719ab8d7c1d6cbc72cd95a0c

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: Preserve func.__qualname__ when defined https://github.com/cloudpipe/cloudpickle/commit/14b38a3ab5970d96cce1492c790494932285f845

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: Some cleanups, fix memoryview support https://github.com/cloudpipe/cloudpickle/commit/f8187e90aed7e1b96ffaae85cdf4b37108c75d3f

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: support for Python 2.6 removed

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: Remove save_reduce() override:  It is the exactly the same code as in Python 2's Pickler class.
https://github.com/cloudpipe/cloudpickle/commit/2da4c243ceddebbc2febf116eb6e53035fed9b9a

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: Further cleanups https://github.com/cloudpipe/cloudpickle/commit/c91aaf110441991307f5097f950764079d0f9652

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: Close StringIO timely on exception https://github.com/cloudpipe/cloudpickle/commit/ca4661b3a20b635f4c240ef763f5759267d74cb9

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: Restore compatibility with functions pickled with 0.4.0 (#128) 
https://github.com/cloudpipe/cloudpickle/commit/7d8c670b703a683d6fd7e642c6bec8a487594d20

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: Preserve func.__qualname__ when defined https://github.com/cloudpipe/cloudpickle/commit/14b38a3ab5970d96cce1492c790494932285f845

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: This didn't seem necessary anymore after the fix for namedtuples

**Comment by BryanCutler** [comment]
@holdenk @HyukjinKwon it seemed like mostly straightforward fixes/cleanups to match cloudpickle 0.4.2 but you two are way more experienced here than me.  Are there any concerns over these updates or additional tests to run?

I did test that namedtuple pickling works with the new fix in cloudpickle, but since the standard pickle still fails we still need the hijack workaround in Spark.

**Comment by holdenk** [comment]
Thank you for going through and documentation the related changes back to the cloudpickle changes :)

File: python/pyspark/cloudpickle.py
Comment (holdenk) [reply python/pyspark/cloudpickle.py]: I'm glad this is moved, should make the next update easier.

**Comment by HyukjinKwon** [comment]
Whoa, nice efforts! Will take a close look within few days.

File: python/pyspark/cloudpickle.py
Comment (holdenk) [reply python/pyspark/cloudpickle.py]: So without the new line between these my brain doesn't parse this right on the first read. What do you think of adding a new line here + back to cloud pickle (can be a follow up)?

File: python/pyspark/cloudpickle.py
Comment (holdenk) [inline python/pyspark/cloudpickle.py]: Looks reasonable since Py 3.4+ only anyways.

**Review by holdenk — COMMENTED** [review]
Thanks for doing this! Not a full review but a quick pass before my talk with minor feedback. I'll look through more later.

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [reply python/pyspark/cloudpickle.py]: so you mean change to this

```
def save_memoryview(self, obj):
        self.save(obj.tobytes())

dispatch[memoryview] = save_memoryview
```

That format is done in a few places and I saw a couple other formatting issues.. so yeah I can submit a PR to do those

**Comment by holdenk** [comment]
So any reason for the WIP tag?

**Comment by BryanCutler** [comment]
I wasn't sure if the named tuple hijack issue from https://issues.apache.org/jira/browse/SPARK-22674 could be fixed here, but it looks like that would require more outside of the scope of this since the problem is with the standard pickling too, right?

**Comment by holdenk** [comment]
Wait, so we left out cloudpickle#113 even though its in 0.4.2?

**Comment by holdenk** [comment]
hmm sorry nvm. So not for this time, but maybe next time we could also copy the cloudpickle_test file over as well.

**Comment by HyukjinKwon** [comment]
> the problem is with the standard pickling too, right?

Yup.

**Comment by BryanCutler** [comment]
> Wait, so we left out cloudpickle#113 even though its in 0.4.2?

That patch is in here and this exactly matches 0.4.2.  I also manually verified that cloudpickle will pickle named tuples with and without the hijack in `types.py`

**Comment by BryanCutler** [comment]
@holdenk and @HyukjinKwon , is there any further testing you guys can think that needs to be done to verify this is ok?

File: python/pyspark/cloudpickle.py
Comment (HyukjinKwon) [reply python/pyspark/cloudpickle.py]: Yea, I think we need this.

**Comment by HyukjinKwon** [comment]
I took a quick look for the commits and seems we should backport https://github.com/cloudpipe/cloudpickle/pull/145 too as looks introduced from https://github.com/cloudpipe/cloudpickle/pull/113. Let me try to backport it to cloudpickle and let's hear their opinion, if I didn't misunderstand.

**Comment by HyukjinKwon** [comment]
LGTM otherwise. @ueshin, I assume you were following this too. Did you had a change to take a look?

Also, @rgbkrk, I think it would be great to have your look too .. It's basically a followup for your PR before.

File: python/pyspark/cloudpickle.py
Comment (rgbkrk) [reply python/pyspark/cloudpickle.py]: That's more of an issue if using pickles stored on disk or if nodes in the cluster are on different versions. Is that likely for Spark?

**Comment by rgbkrk** [comment]
Cool, I can take a gander at this tomorrow (beyond my limited skim just now).

File: python/pyspark/cloudpickle.py
Comment (HyukjinKwon) [reply python/pyspark/cloudpickle.py]: I don't think we made the guarantee on it but the best is always to stay safer. Others seem bug fixes but this one could be a regression (about the support we haven't guaranteed). It's the part of 0.4.2v anyway.

**Comment by ueshin** [comment]
We should also backport https://github.com/cloudpipe/cloudpickle/pull/132? They say that it fixes the regression that affects 0.4.2 and later.

**Comment by HyukjinKwon** [comment]
Nice catch! Yup, I think we should do it too.

**Comment by HyukjinKwon** [comment]
FYI, I manually tried to port cloudpickle#132 and cloudpickle#145 only with corresponding test cases, and then check they were passed.

**Comment by BryanCutler** [comment]
Thanks @ueshin  and @HyukjinKwon , those fixes look reasonable to backport so sounds good to me.  I'll run some tests too and then add to this PR.

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: [MRG] Regression on pickling classes from the main module cloudpipe/cloudpickle#132

File: python/pyspark/cloudpickle.py
Comment (BryanCutler) [inline python/pyspark/cloudpickle.py]: BACKPORTED
BUG: Handle instance methods of builtin types https://github.com/cloudpipe/cloudpickle/pull/145

**Comment by BryanCutler** [comment]
I ran cloudpickle_tests locally and verified that the tests pass with the 2 backported fixes applied.  I did get an unrelated test failure but that happened even before added the fixes and appears to be environment related.

**Comment by HyukjinKwon** [comment]
Yup, I double checked that ahead by manual port. I think that failure might be related with environment which I think is the same I met. If we run the comments as written in the travis, seems passing all fine.

Let me leave my LGTM but let's wait for other looks from other reviewers here.

**Comment by HyukjinKwon** [comment]
Wait .. @BryanCutler, did you port the formatting one here ..? I was thinking we should match it to v0.4.2 as same as possible to reduce the diff.

**Comment by ueshin** [comment]
What about https://github.com/cloudpipe/cloudpickle/pull/140?

**Comment by BryanCutler** [comment]
>Wait .. @BryanCutler, did you port the formatting one here ..? I was thinking we should match it to v0.4.2 as same as possible to reduce the diff.

Yes I added that here as well because I thought that's what Holden asked for, could you confirm @holdenk ?

**Comment by HyukjinKwon** [comment]
@ueshin, is https://github.com/cloudpipe/cloudpickle/pull/140 a regression comparing to our cloudcpikle copy, or do you suggest to port it too as it's a simple and clean bug fix? Seems the same codes were already there.

**Comment by HyukjinKwon** [comment]
Hm .. I was thinking we should focus on matching it to v0.4.2 as same as we can. Few bug links I and @ueshin found looked regressions comparing to our copy so I suggested to include them here ..

**Comment by ueshin** [comment]
@HyukjinKwon Yes, I meant to port it too. I agree with matching it to v0.4.2 as same as we can, so I don't think we should include formatting one but bug fixes.

**Comment by HyukjinKwon** [comment]
Ah, yup. I am okay with including it. Wanted to double check. Seems a small clean bug fix.

**Comment by BryanCutler** [comment]
Yeah the formatting changes will just make it more complicated for the next update, so they should probably not be included now.  cloudpickle#140 looks pretty clean so I'll put that in.

**Comment by BryanCutler** [comment]
I also ran the cloudpickle_tests again after adding cloudpickle#140 (and test) and it is still passing

**Comment by HyukjinKwon** [comment]
retest this please

**Review by rgbkrk — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Will merge this one in few days if there's no more comments.

**Comment by felixcheung** [comment]
this is targeting master, right?

**Comment by HyukjinKwon** [comment]
To me, yup.

**Comment by HyukjinKwon** [comment]
FYI, I am trying to make a minor release of cloudpickle to match with this to deduplicate our efforts. We put many efforts to find and backport bug fixes here. :-).

**Comment by BryanCutler** [comment]
@HyukjinKwon would it be good to update this PR to match the upcoming 4.3 release you are working on? If the code is the same, then just updating the title/description so it is clear

**Comment by HyukjinKwon** [comment]
Yup, now the codes of branch "0.4.x" in cloudpickle is the same with the current PR. Was thinking of letting you know after 0.4.3. Please give me few days ... :-).

**Comment by BryanCutler** [comment]
Sounds good! No rush, I'll keep an eye out for the release

**Comment by HyukjinKwon** [comment]
@BryanCutler, I just released 0.4.3 - https://github.com/cloudpipe/cloudpickle/releases/tag/v0.4.3. Would you mind if I ask to fix PR and JIRA accordingly?

**Comment by BryanCutler** [comment]
Great, thanks @HyukjinKwon!  The 0.4.3 code matches this exactly, so I will just adjust the descriptions.

**Comment by HyukjinKwon** [comment]
I think it's fine in cloudpickle but Spark has the hijacking for regular pickling. I was thinking of a possibility for a deduplicated fix but might have to be investigated separately.

Let's hold this on a bit until the release of 2.3.0 as it's going to go into master anyway (I think). Seems it's been delayed unexpectedly and we better keep the diff small between master and branch-2.3 for now. Will keep my eyes on this PR anyway.

**Comment by BryanCutler** [comment]
>Does the hijacking of the namedtuple still cause problems on Python 3.6?

I'm not too familiar with the history of this, but I ran PySpark tests that cover namedtuples with 3.6.3 and all passed.

**Comment by holdenk** [comment]
So it looks like the 2.3 release is probably going to go out but Jenkins thinks this can't be merged with master. So lets do a jenkins retest this please and I'll try and take some review cycles this week :)

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@holdenk, have you had a chance to take a look for this one?

**Comment by BryanCutler** [comment]
Let me merge with master since it has been sitting a while

**Comment by HyukjinKwon** [comment]
LGTM

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by BryanCutler** [comment]
Thanks @HyukjinKwon @holdenk and @rgbkrk !

**Comment by rgbkrk** [comment]
Woohoo!

**Pull request closed without merging**
